src/vm/runners/builtin_runner.rs

```diff
@@ -132,13 +156,110 @@ impl BuiltinRunner for OutputBuiltinRunner {
     ) -> Option<Vec<MaybeRelocatable>> {
         None
     }
+
+    fn deduce_memory_cell(
+        &mut self,
+        _address: &MaybeRelocatable,
+        _memory: &Memory,
+    ) -> Option<MaybeRelocatable> {
+        None
+    }
+}
+
+#[allow(dead_code)]
+impl HashBuiltinRunner {
+    pub fn new(included: bool, ratio: usize) -> Self {
+        HashBuiltinRunner {
+            base: None,
+            included,
+            _ratio: ratio,
+            _cells_per_instance: 3,
+            _n_input_cells: 2,
+            _stop_ptr: None,
+            verified_addresses: Vec::new(),
+        }
+    }
+}
+
+impl BuiltinRunner for HashBuiltinRunner {
+    fn initialize_segments(&mut self, segments: &mut MemorySegmentManager, memory: &mut Memory) {
+        self.base = Some(segments.add(memory, None))
+    }
+
+    fn initial_stack(&self) -> Vec<MaybeRelocatable> {
+        if self.included {
+            if let Some(builtin_base) = &self.base {
+                vec![MaybeRelocatable::RelocatableValue(builtin_base.clone())]
+            } else {
+                panic!("Uninitialized self.base")
+            }
+        } else {
+            Vec::new()
+        }
+    }
+
+    fn base(&self) -> Option<Relocatable> {
+        self.base.clone()
+    }
+    fn validate_existing_memory(
+        &self,
+        _memory: &[Option<MaybeRelocatable>],
+    ) -> Option<Vec<MaybeRelocatable>> {
+        None
+    }
+    fn deduce_memory_cell(
+        &mut self,
+        address: &MaybeRelocatable,
+        memory: &Memory,
+    ) -> Option<MaybeRelocatable> {
+        if let &MaybeRelocatable::RelocatableValue(ref relocatable) = address {
+            if relocatable.offset % self._cells_per_instance != 2
+                || self.verified_addresses.contains(address)
+            {
+                return None;
+            };
+            if let (Some(MaybeRelocatable::Int(num_a)), Some(MaybeRelocatable::Int(num_b))) = (
+                memory.get(&MaybeRelocatable::RelocatableValue(Relocatable {
+                    segment_index: relocatable.segment_index,
+                    offset: relocatable.offset - 1,
+                })),
+                memory.get(&MaybeRelocatable::RelocatableValue(Relocatable {
+                    segment_index: relocatable.segment_index,
+                    offset: relocatable.offset - 2,
+                })),
+            ) {
+                self.verified_addresses.push(address.clone());
+
+                //Convert MaybeRelocatable to FieldElement
+                let a_string = num_a.to_str_radix(10);
+                let b_string = num_b.to_str_radix(10);
+                let y = match FieldElement::from_dec_str(&a_string) {
```

Comment:
here you can use pattern matching to match the tuple (x,y) the 2 matches. Also it makes sense to return a Result instead of panicking.